pkg/registry/pod/etcd/etcd.go

```diff
@@ -191,10 +205,23 @@ func (r *BindingREST) assignPod(ctx api.Context, podID string, machine string) e
 		return boundPodList, nil
 	})
 	if err != nil {
-		// Put the pod's host back the way it was. This is a terrible hack, but
-		// can't really be helped, since there's not really a way to do atomic
-		// multi-object changes in etcd.
-		if _, err2 := r.setPodHostTo(ctx, podID, machine, ""); err2 != nil {
+		glog.V(1).Infof("failing pod %v because constraint checks failed", podID)
+		podKey, err2 := r.store.KeyFunc(ctx, podID)
+		if err2 == nil {
+			err2 = r.store.Helper.AtomicUpdate(podKey, &api.Pod{}, false, func(obj runtime.Object) (runtime.Object, error) {
+				if pod, ok := obj.(*api.Pod); !ok {
+					return nil, fmt.Errorf("unexpected object: %#v", obj)
+				} else if pod.Status.Host != machine {
```

Comment:
Can you remove the else's here?  Since you're returning they're not necessary and it fits our style more (if the if block returns, never use else).